packages/cli/src/modules/insights/insights-pruning.service.ts

```diff
@@ -0,0 +1,90 @@
+import { Service } from '@n8n/di';
+import { Logger } from 'n8n-core';
+
+import { Time } from '@/constants';
+
+import { InsightsByPeriodRepository } from './database/repositories/insights-by-period.repository';
+import { InsightsConfig } from './insights.config';
+
+@Service()
+export class InsightsPruningService {
+	private pruneInsightsTimer: NodeJS.Timeout | undefined;
+
+	private isStopped = false;
+
+	private readonly delayBetweenRetries = Time.seconds.toMilliseconds * 10;
+
+	private readonly maxRetries = 3;
+
+	constructor(
+		private readonly insightsByPeriodRepository: InsightsByPeriodRepository,
+		private readonly config: InsightsConfig,
+		private readonly logger: Logger,
+	) {
+		this.logger = this.logger.scoped('insights');
+	}
+
+	get isPruningEnabled() {
+		return this.config.maxAgeDays > -1;
+	}
+
+	startPruningTimer() {
+		if (!this.isPruningEnabled) {
+			return;
+		}
+
+		this.clearPruningTimer();
+		this.isStopped = false;
+		this.scheduleNextPrune();
+		this.logger.debug(`Insights pruning every ${this.config.pruneCheckIntervalHours} hours`);
+	}
+
+	private clearPruningTimer() {
+		if (this.pruneInsightsTimer !== undefined) {
+			clearTimeout(this.pruneInsightsTimer);
+			this.pruneInsightsTimer = undefined;
+		}
+	}
+
+	stopPruningTimer() {
+		this.isStopped = true;
+		this.clearPruningTimer();
+		this.logger.debug('Stopped Insights pruning');
+	}
+
+	private scheduleNextPrune() {
+		if (this.isStopped) return;
+
+		this.pruneInsightsTimer = setTimeout(async () => {
+			await this.pruneInsights();
+			this.scheduleNextPrune(); // reschedule after execution
+		}, this.config.pruneCheckIntervalHours * Time.hours.toMilliseconds);
+	}
+
+	async pruneInsights() {
+		for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
+			try {
+				const result = await this.insightsByPeriodRepository.pruneOldData(this.config.maxAgeDays);
```

Comment:
Executions pruning sets a batch size. This is not a concern for insights by period? We expect every cycle to be a small enough deletion?

Reply:
Yes, it should be some order of magnitude less then the pruning of execution data. it will run every 24h by default, and contain around 3 rows by workflow per day max (since at pruning time, granularity will probably be daily or weekly)